pyriemann/utils/covariance.py

```diff
@@ -28,90 +83,98 @@ def _fpcm(X, *, init=None, tol=10e-3, n_iter_max=50, assume_centered=False):
+        and Applications
+        <https://www.researchgate.net/profile/H-Vincent-Poor/publication/258658018_Complex_Elliptically_Symmetric_Distributions_Survey_New_Results_and_Applications/links/550480100cf24cee3a0150e2/Complex-Elliptically-Symmetric-Distributions-Survey-New-Results-and-Applications.pdf>`_
+        E. Ollila, D.E. Tyler, V. Koivunen, H.V. Poor. IEEE Transactions on
+        Signal Processing, 2012.
+    .. [2] `Robust antenna array processing using M-estimators of
+        pseudo-covariance
+        <http://lib.tkk.fi/Diss/2010/isbn9789526030319/article5.pdf>`_
+        E. Ollila, V. Koivunen. PIMRC, 2003.
+    .. [3] `Influence functions for array covariance matrix estimators
+        <https://ieeexplore.ieee.org/abstract/document/1289447/>`_
+        E. Ollila, V. Koivunen. IEEE SSP, 2003.
+    .. [4] `A distribution-free M-estimator of multivariate scatter
         <https://projecteuclid.org/journals/annals-of-statistics/volume-15/issue-1/A-Distribution-Free-M-Estimator-of-Multivariate-Scatter/10.1214/aos/1176350263.full>`_
-        D. E. Tyler. The Annals of Statistics, 1987.
-    .. [2] `Theoretical analysis of an improved covariance matrix estimator in
-        non-Gaussian noise
-        <https://hal.science/hal-02495012/document>`_
-        F. Pascal, P. Forster, J.P. Ovarlez, P. Arzabal. IEEE ICASSP, 2005.
-    .. [3] `Covariance structure maximum-likelihood estimates in compound
-        Gaussian noise: Existence and algorithm analysis
-        <https://hal.science/hal-01816367/document>`_
-        F. Pascal, Y. Chitour, J.P. Ovarlez, P. Forster, P. Arzabal. IEEE
-        Transactions on Signal Processing, 2008.
+        D.E. Tyler. The Annals of Statistics, 1987.
     """  # noqa
     n_channels, n_times = X.shape
+
+    if m_estimator == 'tyl':
+        def weight_func(x):  # Example 2, Section V-C in [1]
+            return n_channels / np.where(x < 1e-10, 1e-10, x)
+    elif m_estimator == 'hub':
+        if not 0 < q <= 1:
+            raise ValueError(f"Value q must be included in (0, 1] (Got {q})")
+
+        def weight_func(x):  # Example 1, Section V-C
+            c2 = chi2.ppf(q, n_channels) / 2
+            b = chi2.cdf(2*c2, n_channels + 1) + c2 * (1-q) / n_channels
+            return np.minimum(1, c2 / x) / b
+    elif m_estimator == 'stu':
+        if nu <= 0:
+            raise ValueError(f"Value nu must be strictly positive (Got {nu})")
+
+        def weight_func(x):  # Eq.(42)
+            return (2*n_channels + nu) / (nu + 2*x)
+    else:
+        raise ValueError(f"Unsupported m_estimator: {m_estimator}")
+
     if not assume_centered:
         X -= np.mean(X, axis=1, keepdims=True)
     if init is None:
-        cov = n_channels * _scm(X, assume_centered=True)
+        cov = _scm(X, assume_centered=True)
     else:
         cov = init
+    X_H = X.conj().T
 
     for _ in range(n_iter_max):
-        diag_ = np.diag(X.T @ np.linalg.inv(cov) @ X)
```

Comment:
This line is not time and memory efficient: As it has to compute a matrix of size (n_times,n_times) to then takes its diag part, it doesn't scale well.

Attached here are two implementations, one I used in robuststats and another one from @flbouchard where he uses a Cholesky decomposition and Lapack inversion. This an important issue because we do this computation multiple times through the fixed-point iteration.

```python
import numpy as np
import numpy.linalg as la
from scipy.linalg.lapack import dtrtri
from scipy.stats import multivariate_normal
from scipy.linalg import toeplitz
from sklearn.covariance import empirical_covariance
from pyriemann.utils.base import invsqrtm
from numpy.testing import assert_almost_equal


n_channels = 500
n_samples = 10000
cov = toeplitz(np.power(0.7, np.arange(0, n_channels)))
X = multivariate_normal.rvs(cov=cov, size=n_samples).T
init = empirical_covariance(X.T)

def weight_func(x):  # Example 2, Section V-C in [1]
    return n_channels / np.where(x < 1e-10, 1e-10, x)

def iteration_diag(X, init):
    X_H = X.conj().T
    powers = np.diag(X_H @ np.linalg.inv(init) @ X)
    cov_new = np.einsum('ab,bc->ac', X * weight_func(powers), X_H)
    return cov_new


def iteration_invsqrtm(X, init):
    temp = invsqrtm(init)@X
    quadratic = np.einsum('ij,ji->i', temp.T, temp)
    quadratic = weight_func(quadratic)
    temp = X * np.sqrt(quadratic)
    return temp@temp.T

def iteration_lapack(X, init):
    L = la.cholesky(init)
    iL, info = dtrtri(L, lower=1)
    v = iL @ X
    a = np.einsum('ij,ji->i', v.T, v)
    Y = np.sqrt(weight_func(a)) * X
    return Y @ Y.T

assert_almost_equal(iteration_diag(X, init), iteration_invsqrtm(X, init))
assert_almost_equal(iteration_diag(X, init), iteration_lapack(X, init))
```
<img width="1063" alt="image" src="https://user-images.githubusercontent.com/44167796/216652864-378c0a71-a36c-4704-b5e8-98ce961c389e.png">

Reply:
Thx @AmmarMian for this valuable suggestion!

Reply:
On complex arrays, `dtrtri` raises this warning:
`ComplexWarning: Casting complex values to real discards the imaginary part`

Reply:
Ah yes probably only work for real-valued data. I'll circle back with Florent who knows well this library if there is a possibility for complex-valued data.

For now maybe the invsqrtm implementation is the best one for all kind of data since it doesn't requires the diag.

Reply:
Actually, we can just use the ctrtri in the complex case:

https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.lapack.ctrtri.html

Reply:
`ctrtri` is ok for complex arrays. However, after several iterations, this implementation generates NaN values and breaks many tests.

Consequently, I use `invsqrtm`-based implementation, more stable and faster than naive one.
I put this code in a specific function `distance_mahalanobis`, and I add a test to compare results with SciPy in the case real-valued data.